capstone/fitness/src/main/java/com/google/sps/util/Scheduler.java

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. 
+package com.google.sps.util;
+
+import com.google.api.client.util.DateTime;
+import com.google.api.services.calendar.model.EventDateTime;
+import com.google.api.services.calendar.model.Event;
+import java.util.*;
+
+// class scheduler uses the user's free/busy information and the duration of an exercise to find a time in a user's day when
+// the user could do an exercise
+public class Scheduler {
+//   private final ScheduleAdapter scheduleAdapter;
+//   private final CalendarAdapter calendarAdapter;
+  private final long exerciseDuration;
+
+
+  public Scheduler(/*ScheduleAdapter scheduleAdapter, CalendarAdapter calendarAdapter,*/ long exerciseDuration) {
+    // this.scheduleAdapter = scheduleAdapter;
+    // this.calendarAdapter = calendarAdapter;
+    this.exerciseDuration =exerciseDuration;
+  }
+ 
+  // minSpan and maxSpan are the boundaries for when a workout could be scheduled on a particular day e.g 7:00 AM to 9:00 PM
+  // events is the user's free/busy information.
+  public Event GetFreeTime(/*EventDateTime day,*/ EventDateTime minSpan, EventDateTime maxSpan, Collection<Event> events) {
+    
+    // initial capacity of eventsQueue is 1 larger than the size of events so that it can handle the case
+    // where there are no events. 
+    PriorityQueue<Event> eventQueue = new PriorityQueue<Event>((events.size() +1), new EventComparator());
+    for (Event evt : events){
+        eventQueue.add(evt);
+    }
+    
+    EventDateTime now = minSpan;
+    EventDateTime end = maxSpan;
+
+
+    long exerciseMilliseconds = this.exerciseDuration * Time.minutesToMilliseconds;
+    
+    // if there is no space for an exercise between 'now' and the next event, then dequeue the first event.
+    while (eventQueue.size() > 0 && Time.eventDateTimeToMilliseconds(eventQueue.peek().getStart()) - Time.eventDateTimeToMilliseconds(now) < (exerciseMilliseconds)) {
```

Comment:
2. Okay! I think the good thing about the freebusy API query is that it ends back the times when the queried user is "busy," so in the case you gave it would just say that the user is busy from 1PM to 4PM, since the events overlap. I will make some more test cases for different scenarios of overlapping events and stuff like that and add them to the design doc. 

1. In the case of not accepted events, I think we can handle them as non-existent since the user could always just accept them if they actually wanted to go? That's my sense, but I'll bring it up in the podsync today to see what Gabriel and Ikenna think.